src/renderers/gl/renderer.cpp

```diff
@@ -642,6 +859,43 @@ void mrg::Renderer::set_output_transform(glm::mat2 const& t)
     }
 }
 
+// Shader that converts colors to grayscale.
+const GLchar* grayscale_src =
+    "uniform sampler2D tex;\n"
+    "vec4 sample_to_rgba(in vec2 texcoord) {\n"
+    "   vec4 col = texture2D(tex, texcoord);\n"
+    "   float s = (col[0] + col[1] + col[2]) / 3.0;\n"
+    "   return vec4(s, s, s, col[3]);\n"
+    "}\n";
+
+// Shader that inverts colors.
+const GLchar* invert_src =
+    "uniform sampler2D tex;\n"
+    "vec4 sample_to_rgba(in vec2 texcoord) {\n"
+    "   vec4 col = texture2D(tex, texcoord);\n"
+    "   return vec4(1.0 - col[0], 1.0 - col[1], 1.0 - col[2], col[3]);\n"
+    "}\n";
+
+void mrg::Renderer::set_output_filter(std::shared_ptr<mg::OutputFilter> const& filter)
+{
+    GLchar const * filter_src;
+    auto filter_name = filter->filter();
+    if (filter_name == "grayscale")
+    {
+        filter_src = grayscale_src;
+    }
+    else if (filter_name == "invert")
+    {
+        filter_src = invert_src;
+    }
+    else
+    {
+        output_filter_shader = nullptr;
+        return;
+    }
+    output_filter_shader = std::make_unique<OutputFilterShader>(output_surface->size().width.as_value(), output_surface->size().height.as_value(), filter_src);
```

Comment:
This is both a threadsafety and `EGLContext`-safety problem; this is *definitely* not going to be called on the compositor thread, so this access races, and we're not guaranteed to have an `EGLContext` current on whatever thread is calling `set_output_filter`.

(Both these problems go away if we move the configuration to `DisplayConfigurationOutput`, because it becomes initialise-only config)

Reply:
I now compile this in bind(), so I think it is safer? Though the changing of the filter might still need some synchronization?